File: mobile/src/main/java/com/alexstyl/specialdates/addevent/AddEventViewModelFactory.java
package com.alexstyl.specialdates.addevent;

import android.view.View;

import com.alexstyl.resources.StringResources;
import com.alexstyl.specialdates.Optional;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.peopleevents.CustomEventType;
import com.alexstyl.specialdates.events.peopleevents.EventType;
import com.alexstyl.specialdates.events.peopleevents.StandardEventType;

import java.util.ArrayList;
import java.util.List;

final class AddEventViewModelFactory {

    private static final Optional<Date> NO_DATE = Optional.absent();
    private final StringResources stringResources;

    AddEventViewModelFactory(StringResources stringResources) {
        this.stringResources = stringResources;
    }

    List<ContactEventViewModel> createViewModelsForAllEventsBut(List<EventType> existingTypes) {
        List<ContactEventViewModel> addEventViewModels = new ArrayList<>();
        boolean containsCustomType = containsCustomType(existingTypes);

        for (StandardEventType eventType : StandardEventType.values()) {
            if (existingTypes.contains(eventType)
                    || eventType == StandardEventType.NAMEDAY
                    || (eventType == StandardEventType.CUSTOM && containsCustomType)
                    ) {
                continue;
            }
            addEventViewModels.add(createAddEventViewModelsFor(eventType));
        }
        return addEventViewModels;
    }

    ContactEventViewModel createAddEventViewModelsFor(EventType eventType) {
        String eventName = eventType.getEventName(stringResources);
        return new ContactEventViewModel(eventName, eventType, NO_DATE, View.GONE);
    }

    private boolean containsCustomType(List<EventType> existingTypes) {
        for (EventType existingType : existingTypes) {
            if (existingType instanceof CustomEventType) {
                return true;
            }
        }
        return false;
    }

}

File: memento/src/main/java/com/alexstyl/specialdates/addevent/AddEventViewModelFactory.kt
package com.alexstyl.specialdates.addevent

import com.alexstyl.specialdates.Optional
import com.alexstyl.specialdates.Strings
import com.alexstyl.specialdates.date.ContactEvent
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.date.DateLabelCreator
import com.alexstyl.specialdates.events.peopleevents.EventType

class AddEventViewModelFactory(private val creator: DateLabelCreator,
                               private val strings: Strings,
                               private val eventIcons: EventIcons) {
    companion object {
        private val NO_DATE = Optional.absent<Date>()
    }

    fun createViewModelFor(contactEvent: ContactEvent): AddEventContactEventViewModel {
        return createViewModelFor(contactEvent.type, contactEvent.date)
    }

    fun createViewModelFor(eventType: EventType, date: Date): AddEventContactEventViewModel {
        val eventHint = creator.createWithYearPreferred(date)
        val dateOptional = Optional(date)
        val clearVisibility = true
        return AddEventContactEventViewModel(eventHint, eventType, dateOptional, clearVisibility, eventIcons.iconOf(eventType))
    }

    fun createViewModelFor(eventType: EventType): AddEventContactEventViewModel {
        val eventName = eventType.getEventName(strings)
        return AddEventContactEventViewModel(eventName, eventType, NO_DATE, false, eventIcons.iconOf(eventType))
    }
}

